### ISSUE TITLE
Rollup: forward-mode AD operator coverage

### ISSUE THREAD

**Issue by soulitzer** [pytorch/pytorch#71117]
These are ops that currently have `supports_forward_ad=False` (usually implicitly) in their OpInfo. Because not all ops have OpInfos yet, some ops that don't support forward AD may be missing from this list: see a formula-oriented list below:

<details>
<summary>
[click to expand]
</summary>

```
_sparse_addmm
affine_grid_generator
acosh_
asinh_
atanh_
atan2
cholesky
cholesky_inverse
clamp.Tensor
_coalesce
polar
logcumsumexp
_ctc_loss
_det_lu_based_helper
dist
_fused_dropout
native_dropout
native_dropout_backward
eig
fake_quantize_per_tensor_affine_cachemask
_fake_quantize_per_tensor_affine_cachemask_tensor_qparams
_fake_quantize_learnable_per_tensor_affine
fake_quantize_per_channel_affine_cachemask
_fake_quantize_learnable_per_channel_affine
_fused_moving_avg_obs_fq_helper
geqrf
grid_sampler_2d
grid_sampler_3d
_grid_sampler_2d_cpu_fallback
igamma
igammac
index.Tensor
special_zeta
special_zeta.self_scalar
special_zeta.other_scalar
logdet
logsumexp
lstsq
lu_unpack
amax
amin
native_batch_norm_backward
native_layer_norm_backward
nextafter
norm.Scalar
norm.ScalarOpt_dim
norm.ScalarOpt_dtype
norm.ScalarOpt_dim_dtype
linalg_vector_norm
_pdist_forward
_pdist_backward
_euclidean_dist
_cdist_forward
_cdist_backward
ormqr
put_
renorm
slogdet
linalg_slogdet
solve
std_mean.correction
rsub.Tensor
rsub.Scalar
nansum
nansum.dim_IntList
_svd_helper
symeig
one_hot
to_dense
to_sparse
to_sparse.sparse_dim
to_mkldnn
_unique
unique_dim
unique_consecutive
unique_dim_consecutive
_unique2
var_mean.correction
_weight_norm_cuda_interface
sparse_mask
_sparse_coo_tensor_with_dims_and_tensors
_sparse_sum.dim
_standard_gamma
_standard_gamma_grad
values
_trilinear
binary_cross_entropy
binary_cross_entropy_backward
embedding
embedding_dense_backward
_embedding_bag
_embedding_bag_dense_backward
embedding_renorm_
multilabel_margin_loss_forward
nll_loss_forward
nll_loss2d_forward
elu_
celu_
glu
log_sigmoid_forward
_log_softmax
_sparse_log_softmax
prelu
prelu_backward
rrelu_with_noise
rrelu_with_noise_
_softmax
_sparse_softmax
_sparse_sparse_matmul
_upsample_bicubic2d_aa.vec
elu_backward
glu_backward
kl_div_backward
l1_loss_backward
log_sigmoid_backward
_log_softmax_backward_data
leaky_relu_backward
max_unpool2d_backward
mse_loss_backward
nll_loss_backward
nll_loss2d_backward
rrelu_with_noise_backward
smooth_l1_loss_backward
huber_loss_backward
softplus_backward
_softmax_backward_data
soft_margin_loss_backward
_upsample_bicubic2d_aa_backward.vec
_cudnn_ctc_loss
cudnn_grid_sampler
cudnn_affine_grid_generator
cudnn_batch_norm_backward
_cudnn_rnn
_cudnn_rnn_backward
miopen_batch_norm_backward
miopen_rnn
miopen_rnn_backward
mkldnn_linear
mkldnn_max_pool2d
mkldnn_max_pool3d
mkldnn_adaptive_avg_pool2d
_mkldnn_reshape
_fft_r2c
_fft_c2r
_fft_c2c
_thnn_fused_lstm_cell
_thnn_fused_gru_cell
_pack_padded_sequence
segment_reduce
_pin_memory
_test_warn_in_autograd
```

</details>

A follow up tracker can be created to track the completion of any remaining formulas if necessary. Alternatively, any additions to the list below are also welcome.

- [ ] atan2
- [ ] cdist
- [ ] cholesky
- [ ] cholesky_inverse
- [ ] symeig
- [ ] clamp
- [ ] fft.fft
- [ ] fft.fft2
- [ ] fft.fftn
- [ ] fft.hfft
- [ ] fft.hfft2
- [ ] fft.hfftn
- [ ] fft.rfft
- [ ] fft.rfft2
- [ ] fft.rfftn
- [ ] fft.ifft
- [ ] fft.ifft2
- [ ] fft.ifftn
- [ ] fft.ihfft
- [ ] fft.ihfft2
- [ ] fft.ihfftn
- [ ] fft.irfft
- [ ] fft.irfft2
- [ ] fft.irfftn
- [ ] stft
- [ ] istft
- [ ] linalg.det
- [ ] linalg.det
- [ ] linalg.cond
- [ ] linalg.eigvals
- [ ] linalg.lstsq
- [ ] linalg.norm
- [ ] linalg.matrix_norm
- [ ] linalg.slogdet
- [ ] linalg.vector_norm
- [ ] lu_unpack
- [ ] softmax
- [ ] softmax
- [ ] nn.functional.softmin
- [ ] nn.functional.softmin
- [ ] nn.functional.cross_entropy
- [ ] nn.functional.normalize
- [ ] nn.functional.bilinear
- [ ] nn.functional.glu
- [ ] nn.functional.prelu
- [ ] nn.functional.rrelu
- [ ] nn.functional.logsigmoid
- [ ] igamma
- [ ] igammac
- [ ] dist
- [ ] `__rsub__`
- [ ] rsub
- [ ] rsub
- [ ] solve
- [ ] eig
- [ ] svd
- [ ] linalg.svd
- [ ] linalg.svdvals
- [ ] svd_lowrank
- [ ] pca_lowrank
- [ ] polar
- [ ] `__getitem__`
- [ ] put
- [ ] renorm
- [ ] special.zeta
- [ ] logsumexp
- [ ] to_sparse
- [ ] logcumsumexp
- [ ] logdet
- [ ] log_softmax
- [ ] log_softmax
- [ ] norm
- [ ] norm
- [ ] norm
- [ ] norm
- [ ] nn.functional.one_hot
- [ ] nn.functional.embedding
- [ ] nn.functional.embedding_bag
- [ ] nn.functional.grid_sample
- [ ] amax
- [ ] amin
- [ ] nanmean
- [ ] nansum
- [ ] _masked.amax
- [ ] _masked.amin
- [ ] _masked.norm
- [ ] _masked.softmax
- [ ] _masked.log_softmax
- [ ] _masked.softmin
- [ ] _masked.normalize
- [ ] nn.functional.ctc_loss
- [ ] nn.functional.nll_loss
- [ ] nn.functional.pairwise_distance

See also: Tracker for `supports_fwgrad_bwgrad=False` for forward-over-reverse Hessian coverage (TODO).

**Comment by Lezcano** [comment]
I added a few operations that will have AD support once some PRs are merged, and I also removed a few duplicated functions. It's worth noting that this list includes some operations that will be removed in the next version, such as `symeig` and similar linalg functions.

**Comment by nikitaved** [comment]
Oh, awesome! Thank you! Now we have list of small PRs in-between large things :)
There are some linear algebra functions on this list that are going to be deprecated, so they could get excluded. Most torch.linalg functions support forward AD.

**Comment by soulitzer** [comment]
Updated to exclude the deprecated linear algebra functions that will be deprecated, thanks!

**Comment by nikitaved** [comment]
About linalg: `linalg.lstsq` has forward AD implemented. svd/pca-lowrank should be able to support forward AD as they are composite operations relying on mm and svd, for which `linalg.svd` now has forward AD support, unless legacy svd is used...

**Comment by zou3519** [comment]
NB: @pmeier's old at https://github.com/pytorch/pytorch/pull/69211 has many of the OpInfos that we need to test the above (binary_cross_entropy{_with_logits}, l1_loss, etc) but I don't know what the status of it is

**Comment by pmeier** [comment]
@mruberry wanted to land the stack at some point, but it seems we both have forgotten about it. Want to give this another shot?
